rust/lance/src/dataset.rs

```diff
@@ -486,17 +488,63 @@ impl Dataset {
             });
         }
 
-        let fragments = write_fragments(
-            object_store.clone(),
-            &self.base,
-            &schema,
-            stream,
-            params.clone(),
-        )
-        .await?;
+        let transaction = match (params.tail_compaction_size, self.get_fragments().last()) {
+            (Some(tail_compact_thresh), Some(last_frag))
+                if last_frag.count_rows().await? <= tail_compact_thresh =>
+            {
+                let last_data = last_frag.scan().try_into_stream().await?;
+                let last_frag_stream: SendableRecordBatchStream = last_data.into();
+                let arrow_schema = last_frag_stream.schema();
+                let merged_stream = {
+                    // need to scope this use because TokioStreamExt collides with futures::StreamExt
+                    use tokio_stream::StreamExt as TokioStreamExt;
+                    RecordBatchStreamAdapter::new(arrow_schema, last_frag_stream.merge(stream))
+                };
 
-        let transaction =
-            Transaction::new(self.manifest.version, Operation::Append { fragments }, None);
+                let fragments = write_fragments(
+                    object_store.clone(),
+                    &self.base,
+                    &schema,
+                    Box::pin(merged_stream),
+                    params.clone(),
+                )
+                .await?;
+
+                if fragments.len() > 1 {
+                    return Err(Error::InvalidInput {
+                        source: "When attempting to run tail compaction, ".into(),
+                    });
+                }
+                // maybe we should roll append into a rewrite somehow?
+                // append and rewrite are both remove and/or appending fragments
+
+                // Concurrent writes fails on this path
+                // NOCOMMIT: explain how this interacts with concurrent
```

Comment:
My concern with concurrency with this approach is that once a database has a small trailing fragment then any concurrent writes will suddenly conflict.

For example:

Db trailing fragment is size 200

Incoming write of size X
Incoming write of size Y

These two incoming writes will conflict since they both try and merge in the small trailing fragment.

I think we will eventually need to be able to support concurrent writes in combination with tail-compaction.

Reply:
yeah I agree, this effectively makes concurrent write not possible.

We would need some kind of global synchronization mechanism to serialize the operations here.

Another way I could see this being done is:
1. check if table needs compaction
2. request compaction in a global queue <-- This dedupes all the concurrent compaction requests from different writer
3. Wait for request to finish
4. Insert as usual